File: conformance/tests/stateful/stateful.tsop.go
package contract

import "tsop"

type Stateful struct {
	tsop.StatefulSmartContract
	Count    tsop.Int
	MaxCount tsop.Int `tsop:"readonly"`
}

func (c *Stateful) Increment(amount tsop.Int, txPreimage tsop.SigHashPreimage) {
	tsop.Assert(tsop.CheckPreimage(txPreimage))
	c.Count = c.Count + amount
	tsop.Assert(c.Count <= c.MaxCount)
	tsop.Assert(tsop.Hash256(c.GetStateScript()) == tsop.ExtractOutputHash(txPreimage))
}

func (c *Stateful) Reset(txPreimage tsop.SigHashPreimage) {
	tsop.Assert(tsop.CheckPreimage(txPreimage))
	c.Count = 0
	tsop.Assert(tsop.Hash256(c.GetStateScript()) == tsop.ExtractOutputHash(txPreimage))
}

File: conformance/tests/stateful/stateful.tsop.rs
use tsop::prelude::*;

#[tsop::contract]
struct Stateful {
    count: Int,
    #[readonly]
    max_count: Int,
}

#[tsop::methods(Stateful)]
impl Stateful {
    #[public]
    fn increment(&mut self, amount: Int, tx_preimage: SigHashPreimage) {
        assert!(check_preimage(tx_preimage));
        self.count = self.count + amount;
        assert!(self.count <= self.max_count);
        assert!(hash256(self.get_state_script()) == extract_output_hash(tx_preimage));
    }

    #[public]
    fn reset(&mut self, tx_preimage: SigHashPreimage) {
        assert!(check_preimage(tx_preimage));
        self.count = 0;
        assert!(hash256(self.get_state_script()) == extract_output_hash(tx_preimage));
    }
}
